kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt

```diff
@@ -213,18 +213,14 @@ class KotlinJsonAdapterFactory : JsonAdapter.Factory {
       val parameter = parametersByName[property.name]
 
       if (Modifier.isTransient(property.javaField?.modifiers ?: 0)) {
-        @Suppress("ReplaceGuardClauseWithFunctionCall") // More readable this way
-        if (parameter != null && !parameter.isOptional) {
-          throw IllegalArgumentException(
-              "No default value for transient constructor $parameter")
+        require(parameter == null || parameter.isOptional) {
+          "No default value for transient constructor $parameter"
         }
         continue
       }
 
-      @Suppress("ReplaceGuardClauseWithFunctionCall") // More readable this way
-      if (parameter != null && parameter.type != property.returnType) {
-        throw IllegalArgumentException("'${property.name}' has a constructor parameter of type " +
-            "${parameter.type} but a property of type ${property.returnType}.")
+      require(parameter == null || parameter.type == property.returnType) {
+        "'${property.name}' has a constructor parameter of type ${parameter!!.type} but a property of type ${property.returnType}."
```

Comment:
Compiler can't smartcast this `parameter` here

Reply:
File a bug?

Reply:
Can do, but wondering if this is just a known limitation of contracts for now (e.g. no compound expressions)